src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -0,0 +1,458 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_x509.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+
+#define INIT_LOCALS(name, ...) \
+    enum { __VA_ARGS__, count_##name }; \
+    jobject name[count_##name] = { 0 }; \
+
+#define RELEASE_LOCALS(name, env) \
+{ \
+    for (int i = 0; i < count_##name; ++i) \
+    { \
+        jobject local = name[i]; \
+        if (local != NULL) \
+            (*env)->DeleteLocalRef(env, local); \
+    } \
+} \
+
+#define VALUE_FAIL -1
+#define BUFFER_FAIL FAIL
+
+static int32_t PopulateByteArray(JNIEnv *env, jbyteArray source, uint8_t *dest, int32_t len);
+static int32_t PopulateString(JNIEnv *env, jstring source, char *dest, int32_t len);
+
+// Handles both DER and PEM formats
+jobject /*X509Certificate*/ AndroidCryptoNative_DecodeX509(const uint8_t *buf, int32_t len)
+{
+    assert(buf != NULL && len > 0);
+    JNIEnv* env = GetJNIEnv();
+
+    jobject ret = NULL;
+    INIT_LOCALS(loc, bytes, stream, certType, certFactory)
+
+    // byte[] bytes = new byte[] { ... }
+    // InputStream stream = new ByteArrayInputStream(bytes);
+    loc[bytes] = (*env)->NewByteArray(env, len);
```

Comment:
I've been refraining from asking because it feels like a stupid question... but:

Why do `(*env)` so many times?  Seems like 

```C
JNIEnv* env = GetJNIEnv();
JNIEnv envObj = *env;

...

loc[bytes] = envObj->NewByteArray(env, len);
```

would save a lot of redundant memory dereferences.

Reply:
This is recommended approach for JNI. Dereferencing like this gives the desired late dispatch behaviour in C, such that it can be compatible across JVM versions.

@AaronRobinsonMSFT would be able to give a more detailed explanation offline if you are interested.